processing/src/main/java/io/druid/query/MetricsEmittingQueryRunner.java

```diff
@@ -90,116 +90,50 @@ public MetricsEmittingQueryRunner(
-        }
-        catch (Error e) {
-          builder.setDimension(DruidMetrics.STATUS, "failed");
-          throw e;
-        }
-
-        return makeYielder(startTime, retVal, builder);
-      }
-
-      private <OutType> Yielder<OutType> makeYielder(
-          final long startTime,
-          final Yielder<OutType> yielder,
-          final ServiceMetricEvent.Builder builder
-      )
-      {
-        return new Yielder<OutType>()
+    return Sequences.wrap(
+        // Use LazySequence because want to account execution time of queryRunner.run() (it prepares the underlying
+        // Sequence) as part of the reported query time, i. e. we want to execute queryRunner.run() after
+        // `startTime = System.currentTimeMillis();` (see below).
+        new LazySequence<>(new Supplier<Sequence<T>>()
         {
           @Override
-          public OutType get()
+          public Sequence<T> get()
           {
-            return yielder.get();
-          }
-
-          @Override
-          public Yielder<OutType> next(OutType initValue)
-          {
-            try {
-              return makeYielder(startTime, yielder.next(initValue), builder);
-            }
-            catch (RuntimeException e) {
-              builder.setDimension(DruidMetrics.STATUS, "failed");
-              throw e;
-            }
-            catch (Error e) {
-              builder.setDimension(DruidMetrics.STATUS, "failed");
-              throw e;
-            }
+            return queryRunner.run(query, responseContext);
           }
+        }),
+        new SequenceWrapper()
+        {
+          private long startTime;
 
           @Override
-          public boolean isDone()
+          public void before()
           {
-            return yielder.isDone();
+            startTime = System.currentTimeMillis();
           }
 
           @Override
-          public void close() throws IOException
+          public void after(boolean isDone, Throwable thrown)
           {
             try {
-              if (!isDone() && builder.getDimension(DruidMetrics.STATUS) == null) {
-                builder.setDimension(DruidMetrics.STATUS, "short");
-              }
-
               long timeTaken = System.currentTimeMillis() - startTime;
               emitter.emit(builder.build(metricName, timeTaken));
 
               if (creationTime > 0) {
                 emitter.emit(builder.build("query/wait/time", startTime - creationTime));
               }
             }
+            // Use finally block because emitting query time (in `try {}`, above) and the status (in `finally {}`,
+            // below) are unrelated and we don't want the latter to be skipped if the former thrown any exception.
             finally {
-              yielder.close();
+              if (thrown != null) {
+                builder.setDimension(DruidMetrics.STATUS, "failed");
+              } else if (!isDone) {
+                builder.setDimension(DruidMetrics.STATUS, "short");
```

Comment:
It's sketchy to call `builder.setDimension` here, after `emitter.emit(builder.build(...))` has already been called. This will be going in and editing a hashmap that has already been handed off to the emitter.